### DISCUSSION TITLE
Persist response body outside Handle()

### DISCUSSION THREAD

**Discussion by tmm1** [earthboundkid/requests#94] [Support]
i'm making a request for a large binary blob, and would like to return the underlying req.Body to my caller so they can drain it themselves

i'm using the following:

```
Handle(func(resp *http.Response) error {
	return (*requests.ResponseError)(resp)
})
```

which lets me shuffle the response out as an error. however, due to this line:

https://github.com/carlmjohnson/requests/blob/5d11526188c82fefcbdbde64537a4ad49841cdff/core_do.go#L21

the body is closed, and reading it errors out with: `http: read on closed response body`

i tried adding a `resp.Body = io.NopCloser(resp.Body)` to my handler, but that didn't work.

is there a simple workaround i'm missing?

**Comment by tmm1** [comment]
replacing `req.Body` after the defer call doesn't work: https://go.dev/play/p/AQO3Vzp5b3D

**Reply by earthboundkid** [reply]
The preferred approach is to invert control and receive a callback that takes an io.Reader, so the Body can close when the callback does. If you can’t invert control for some reason, you could spin up a goroutine and have it keep the Handler open until you’re ready. At that point though, you lose a lot of the point of using requests, which is automatic Body closing, since you would need manual goroutine closing instead.
